sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -90,6 +90,117 @@ case class MapKeys(child: Expression)
+  """,
+  since = "2.4.0")
+case class Zip(children: Seq[Expression]) extends Expression with ExpectsInputTypes {
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.length)(ArrayType)
+
+  def mountSchema(): StructType = {
+    val arrayAT = children.map(_.dataType.asInstanceOf[ArrayType])
+    val fields = arrayAT.zipWithIndex.foldRight(List[StructField]()) {
+      (item, list) => {
+        val (arr, idx) = item
+        StructField(s"_$idx", arr.elementType, arr.containsNull) :: list
+      }
+    }
+    StructType(fields)
+  }
+
+  override def dataType: DataType = ArrayType(mountSchema())
+
+  override def prettyName: String = "zip"
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val genericArrayData = classOf[GenericArrayData].getName
+    val genericInternalRow = classOf[GenericInternalRow].getName
+
+    val evals = children.map(_.genCode(ctx))
+    val numArrs = evals.length
+
+    val arrCardinality = ctx.freshName("args")
+    val arrVals = ctx.freshName("arrVals")
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |}
+      """.stripMargin
+    }.mkString("\n")
+
+    val myobject = ctx.freshName("myobject")
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+    val j = ctx.freshName("j")
+    val i = ctx.freshName("i")
+    val args = ctx.freshName("args")
+
+    ev.copy(s"""
+      |ArrayData[] $arrVals = new ArrayData[$numArrs];
+      |int[] $arrCardinality = new int[$numArrs];
+      |$inputs
+      |int $biggestCardinality = 0;
+      |for (int $i = 0; $i < $numArrs; $i ++) {
+      |  $arrCardinality[$i] = $arrVals[$i].numElements();
+      |  $biggestCardinality = Math.max($biggestCardinality, $arrCardinality[$i]);
+      |}
+      |Object[] $args = new Object[$biggestCardinality];
+      |for (int $i = 0; $i < $biggestCardinality; $i ++) {
+      |  Object[] $myobject = new Object[$numArrs];
+      |  for (int $j = 0; $j < $numArrs; $j ++) {
+      |    if ($arrCardinality[$j] > $i) {
+      |      $myobject[$j] = $arrVals[$j].getInt(0);
+      |    } else {
+      |      $myobject[$j] = null;
+      |    }
+      |  }
+      |  $args[$i] = new $genericInternalRow($myobject);
+      |}
+      |$genericArrayData ${ev.value} = new $genericArrayData($args);
+    """.stripMargin)
+  }
+
+  override def nullable: Boolean = children.forall(_.nullable)
+
+  override def eval(input: InternalRow): Any = {
+    val inputArrays = children.map(_.eval(input).asInstanceOf[ArrayData])
+    val arrayTypes = children.map(_.dataType.asInstanceOf[ArrayType].elementType)
+    val numberOfArrays = children.length
+
+    var biggestCardinality = 0
+    for (e <- inputArrays) {
+      biggestCardinality = biggestCardinality max e.numElements()
+    }
+
+    var i = 0
+    var j = 0
+    var result = Seq[InternalRow]()
```

Comment:
You already know the final size. Why not use `new Array[InternalRow](biggestCardinallity)`?